ms2deepscore/spectrum_pair_selection.py

```diff
@@ -0,0 +1,278 @@
+        self._scores = []
+        self._cols = []
+        self.shuffling = shuffling
+        self._idx_to_inchikey = dict(enumerate(inchikeys))
+        self._inchikey_to_idx = {key: idx for idx, key in enumerate(inchikeys)}
+
+        for row_idx in self._idx_to_inchikey.keys():
+            row_mask = (coo_array.row == row_idx)
+            self._cols.append(coo_array.col[row_mask])
+            self._scores.append(coo_array.data[row_mask])
+
+        # Initialize counter for each column
+        self._row_generator_index = np.zeros(len(self._idx_to_inchikey), dtype=int)
+        if self.shuffling:
+            self.shuffle()
+
+    def shuffle(self):
+        """Shuffle all scores for all inchikeys."""
+        for i in range(len(self._scores)):
+            self._shuffle_row(i)
+
+    def _shuffle_row(self, row_index):
+        """Shuffle the column and scores of row with row_index."""
+        permutation = np.random.permutation(len(self._cols[row_index]))
+        self._cols[row_index] = self._cols[row_index][permutation]
+        self._scores[row_index] = self._scores[row_index][permutation]
+
+    def next_pair_for_inchikey(self, inchikey):
+        row_idx = self._inchikey_to_idx[inchikey]
+
+        # Retrieve the next pair
+        col_idx = self._cols[row_idx][self._row_generator_index[row_idx]]
+        score = self._scores[row_idx][self._row_generator_index[row_idx]]
+
+        # Update the counter, wrapping around if necessary
+        self._row_generator_index[row_idx] += 1
+        if self._row_generator_index[row_idx] >= len(self._cols[row_idx]):
+            self._row_generator_index[row_idx] = 0
+            # Went through all scores in this row --> shuffle again
+            self._shuffle_row(row_idx)
+
+        return score, self._idx_to_inchikey[col_idx]
+
+    def generator(self):
+        """Infinite generator to loop through all inchikeys."""
+        while True:
+            for inchikey in self._inchikey_to_idx.keys():
+                score, inchikey2 = self.next_pair_for_inchikey(inchikey)
+                yield inchikey, score, inchikey2
+
+    @property
+    def scores(self):
+        return self._scores
+
+    @property
+    def idx_to_inchikey(self):
+        return self._idx_to_inchikey
+
+    @property
+    def inchikey_to_idx(self):
+        return self._inchikey_to_idx
+
+    def __str__(self):
+        return f"SelectedCompoundPairs with {len(self._scores)} columns."
+
+
+def compute_spectrum_pairs(spectrums,
+                           selection_bins: np.ndarray = np.array([(x/10, x/10 + 0.1) for x in range(0, 10)]),
+                           max_pairs_per_bin: int = 10,
+                           include_diagonal: bool = True,
+                           fix_global_bias: bool = True,
+                           fingerprint_type: str = "daylight",
+                           nbits: int = 2048):
+    """Function to compute the compound similarities (Tanimoto) and collect a well-balanced set of pairs.
```

Comment:
We could in addition make a class that stores all ms2deepscore training settings, to prevent passing this many arguments from function to function and to add a method for printing the used settings (maybe a yaml file?).

This would look like:
```python
class TrainingSettings():
    def __init__(max_pairs_per_bin: int = 10,
                           include_diagonal: bool = True,
                           fix_global_bias: bool = True,
                           fingerprint_type: str = "daylight",
                           neural_net_architecture,
                           pair_generation_function,
                           etc...)
    self.max_pairs_per_bin = max_pairs_per_bin
    etc...

    def store_setttings(file_name):
        # Saves the settings in a yaml file. 

```

Reply:
All these suggestions do not have to implemented in this PR of course, it probably makes more sense to convert them into issues to implement in the future.

Reply:
> I would restructure these functions a bit, to make it a bit more modular. Suggested changes: Make a function, calculate fingerprints. Combine compute_spectrum_pairs and jaccard_similarity_matrix_cherrypiking into one wrapper function. Or maybe not even a wrapper function, but instead make this part of the example code for training a MS2Deepscore model, since this will give the flexibility of using a different pair generation algorithm instead of jaccard_similarity_matrix_cherrypicking (for instance on ionmode).

This has been implemented as wel in #146